cmd/tcplb/flags.go

```diff
@@ -0,0 +1,75 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"net"
+	"strings"
+	"tcplb/lib/core"
+)
+
+const (
+	commandName     = "tcplb"
+	upstreamListSep = ","
+)
+
+var InvalidUpstreamListFlag = errors.New("invalid upstream list flag value")
+
+// UpstreamListValue is a flag.Value for lists of Upstream addresses.
+type UpstreamListValue struct {
+	Upstreams []core.Upstream
+}
+
+func (v *UpstreamListValue) String() string {
+	n := len(v.Upstreams)
+	tokens := make([]string, n, n)
+	for i, u := range v.Upstreams {
+		tokens[i] = u.Address
+	}
+	return strings.Join(tokens, upstreamListSep)
+}
+
+func (v *UpstreamListValue) Set(s string) error {
+	tokens := strings.Split(s, upstreamListSep)
+	for _, token := range tokens {
+		host, port, err := net.SplitHostPort(token)
+		if err != nil {
+			return InvalidUpstreamListFlag
+		}
+		upstream := core.Upstream{
+			Network: defaultUpstreamNetwork,
+			Address: net.JoinHostPort(host, port),
+		}
+		v.Upstreams = append(v.Upstreams, upstream)
+	}
+	return nil
+}
+
+func newConfigFromFlags(argv []string) (*Config, error) {
```

Comment:
It's a bit uncommon to detach flags from where they are used. A bunch of global vars at the top of the file is the more usual idiom (it's a bit weird if you come from other languages, but it's alright).

Reply:
one advantage of doing it in this unidiomatic way is that the config parsing can be unit tested without getting tangled up with global state.

then again, i didnt write any unit tests, so this isnt the most convincing argument